frb_codegen/src/commands.rs

```diff
@@ -18,11 +17,34 @@ fn call_shell(cmd: &str) -> Output {
     execute_command("sh", &["-c", cmd], None)
 }
 
-pub fn ensure_tools_available() -> Result {
-    let output = call_shell("dart pub global list");
-    let output = String::from_utf8_lossy(&output.stdout);
-    if !output.contains("ffigen") {
-        return Err(Error::MissingExe(String::from("ffigen")));
+pub fn ensure_tools_available(dart_root: &str) -> Result {
+    let toolchain = guess_toolchain(dart_root).unwrap();
+    match toolchain {
+        DartToolchain::Dart if !call_shell("dart --version").status.success() => {
+            return Err(Error::MissingExe("dart".to_string()));
+        }
+        DartToolchain::Flutter if !call_shell("flutter --version").status.success() => {
+            return Err(Error::MissingExe("flutter".to_string()));
+        }
+        _ => {}
+    };
+
+    let ffi = ensure_dependencies(dart_root, "ffi", PackageManager::Dependencies).unwrap();
+    if ffi.is_none() {
+        return Err(Error::MissingDep {
+            name: "ffi".into(),
+            context: PackageManager::Dependencies,
+            version: "^2.0.1".into(),
+        });
+    }
```

Comment:
nit: seems that can still extract a function about such dep checking. e.g.

```
fn ensure_dependencies_v2() {
  let success = ...check_do_we_have_dep...;
  if !success {
    return Err(...);
  }
  Ok(())
}

Reply:
since there's the restrictions about semver requirements checking, what do you have in mind here ?
- peek into `pubspec.lock` to make sure downloaded version is in range ?
- check into `pubspec.yaml` that the dependency is specified exactly as required ?

Reply:
Well I guess it is two questions.

1. Pure code-refactoring problem: Notice we still have code duplication across ffi and ffigen checking, so we can extract functions
2. How to check semver

> peek into pubspec.lock to make sure downloaded version is in range ?

Given the difficulties discussed above (e.g. semver `^` has weird behavior), seems this is easier. Since the lock file tells us which *exact* version is there.

Reply:
actually I'd rather leave it for a subsequent PR : I already implemented the changes in a [derived PR](https://github.com/Roms1383/flutter_rust_bridge/pull/1), and there's quite a lot of changes, which makes me wonder if we're not going beyond the scope of the current PR.
what do you think ?

Reply:
Sure, separate changes into multiple PRs can make code review and tracking easier :)

Reply:
ok let's resolve it for now and track it in the subsequent PR then.